# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: js/src/gc/GC.cpp
# Commit: 417d42aa08ef
# Full Hash: 417d42aa08efdd20ba112e3ff1e633338fba111d
# Author: Jon Coppeard <jcoppeard@mozilla.com>
# Date: 2024-07-12 22:06:45
# Description:
#   Bug 1871303 - Ensure there is no more marking at the start of sweeping r=sfink
#   
#   There are two issues here. The first is that there unexpected marking work at
#   the start of sweeping after entering from the mark phase without yielding. We
#   previously called assertNoMarkingWork() after markUntilBudgetExhausted() in the
# ==============================================================================

diff -r 3396a3835684 -r 417d42aa08ef js/src/gc/GC.cpp
--- a/js/src/gc/GC.cpp	Fri Jul 12 09:30:34 2024 +0000
+++ b/js/src/gc/GC.cpp	Fri Jul 12 09:33:13 2024 +0000
@@ -3819,10 +3819,7 @@
 
     case State::Mark:
       if (mightSweepInThisSlice(budget.isUnlimited())) {
-        // Trace wrapper rooters before marking if we might start sweeping in
-        // this slice.
-        rt->mainContextFromOwnThread()->traceWrapperGCRooters(
-            marker().tracer());
+        prepareForSweepSlice(reason);
 
         // Incremental marking validation re-runs all marking non-incrementally,
         // which requires collecting the nursery. If that might happen in this
@@ -3879,13 +3876,8 @@
       [[fallthrough]];
 
     case State::Sweep:
-      if (storeBuffer().mayHavePointersToDeadCells()) {
-        collectNurseryFromMajorGC(reason);
-      }
-
       if (initialState == State::Sweep) {
-        rt->mainContextFromOwnThread()->traceWrapperGCRooters(
-            marker().tracer());
+        prepareForSweepSlice(reason);
       }
 
       if (performSweepActions(budget) == NotFinished) {